Mage/src/main/java/mage/game/command/emblems/TyvarKellEmblem.java

```diff
@@ -0,0 +1,105 @@
+package mage.game.command.emblems;
+
+import mage.MageObjectReference;
+import mage.abilities.Ability;
+import mage.abilities.common.SpellCastControllerTriggeredAbility;
+import mage.abilities.effects.ContinuousEffectImpl;
+import mage.abilities.effects.common.DrawCardSourceControllerEffect;
+import mage.abilities.keyword.HasteAbility;
+import mage.cards.Card;
+import mage.constants.*;
+import mage.filter.FilterSpell;
+import mage.game.Game;
+import mage.game.command.Emblem;
+import mage.game.permanent.Permanent;
+import mage.game.stack.Spell;
+
+import java.util.Iterator;
+
+/**
+ *
+ * @author weirddan455
+ */
+public final class TyvarKellEmblem extends Emblem {
+
+    private static final FilterSpell filter = new FilterSpell("an Elf spell");
+
+    static {
+        filter.add(SubType.ELF.getPredicate());
+    }
+
+    // −6: You get an emblem with "Whenever you cast an Elf spell, it gains haste until end of turn and you draw two cards."
+    public TyvarKellEmblem() {
+        this.setName("Emblem Tyvar");
+        this.setExpansionSetCodeForImage("KHM");
+        Ability ability = new SpellCastControllerTriggeredAbility(Zone.COMMAND, new TyvarKellEmblemEffect(
+                HasteAbility.getInstance(), Duration.EndOfTurn), filter, false, true
+        );
+        ability.addEffect(new DrawCardSourceControllerEffect(2, "you").concatBy("and"));
+        this.getAbilities().add(ability);
+    }
+}
+
+class TyvarKellEmblemEffect extends ContinuousEffectImpl {
```

Comment:
Why don't you use a standard `GainAbilityTargetEffect`?

![shot_210111_014658](https://user-images.githubusercontent.com/8344157/104136351-f1bc5a80-53ae-11eb-9806-8199a5ce67b5.png)

Reply:
I tried that first and it didn't work even with `onCard` set.  That class expects to have a Card ID as the target but it's getting passed a Spell ID so when it checks the card with `Card card = game.getCard(cardId);`, it's null.  It also doesn't apply the effect to the Permanent.

Reply:
Original problem was in gain ability effect, fixed target and zcc. If you give ability to the card on the stack then permanent can't take it after resolve. Reason: original card goes to graveyard and change own zcc, so gain ability is no more affecting it and permanent lost that ability too.

I'll look at it later today -- is it possible to improve standard abilities to support permanent related abilities on the card. If it too hard then your version will be approved. Don't merge that PR.

Reply:
Additional improves added by 6dcbcbe962fdda5d6b0590d13b7c7487605363ec, so current card can use a standard code.